cfpb/cfgov-refresh
# WIP Feature: add line chart block to Wagtail

Looking for feedback on this Wagtail code from @cfpb/cfgov-backends! Check out `GHE/CFGOV/data-research/issues/526` for more context.

This WIP adds a new Chart Block to the Wagtail admin, for entering metadata to generate d3 charts for the Consumer Credit Trends pages.

Content entered in the Chart Block fields will replace the metadata that currently powers the charts in this file: https://github.com/cfpb/consumer-credit-trends/blob/master/src/static/js/templates/charts.js

This will allow our product owner to enter and manage the content for each chart.


## Additions

- Atomic chart organism JavaScript, based on JS from [consumer-credit-trends line.js](https://github.com/cfpb/consumer-credit-trends/blob/master/src/static/js/line.js), including dependencies
- In the spirit of iterative updates, this PR includes the Line Chart code only, for easier review.
- `chart.less` adds styles from  [consumer-credit-trends charts.less](https://github.com/cfpb/consumer-credit-trends/blob/master/src/static/css/charts.less), with updated atomic class names
- New Wagtail block fields required to make the JavaScript render each chart. These include what type of chart to load (this can be used later to determine which JS file to call), and the group property, which is used to filter data for many line charts.

## Removals

- `elementID` field from Chart Block. We don't want to rely on an ID for each chart to render any longer, so this is a step towards that. The JS now expects a matching title + class name instead of the ID. This will require manually updated migration.

## Changes

- `chart.html` template class names have been atomicized to match project conventions
- The original helper functions in the  [consumer-credit-trends line.js](https://github.com/cfpb/consumer-credit-trends/blob/master/src/static/js/line.js) assume that no `chartInfo.group` property is defined for charts that don't need it. Because this updated JS grabs the group property from the HTML data attribute, group property will exist for all charts.
- Another note about groups. Right now the JS expects the following properties for certain types of graphs. We can probably do the data munging/filtering on the Jenkins job that @contolini and @hillaryj are working on for these, instead of in the JS:
  - Origination activity, number/volume line charts: no group
  - Origination activity, year over year number/volume bar charts: "Number of loans" or "Dollar Volume"
  - Borrower Risk Profiles, Line charts: Credit score group (Subprime, prime, etc)
  - Income level line charts: income group (low, moderate, etc)
  - Borrower age line charts:  age group (younger than 30, 30-44, etc)


## Testing

- check out and create a "Chart block" element on a browse page to test the fields display on the page.
- add required fields, use screenshot below and [charts.js properties](https://github.com/cfpb/consumer-credit-trends/blob/master/src/static/js/templates/charts.js) as a guide
- test that multiple line charts can be added to a page
- test that colors of line charts match the report type specified in the CSS

## Screenshots

### Wagtail admin fields

![screen shot 2017-01-14 at 12 13 26 am](https://cloud.githubusercontent.com/assets/702526/21952564/71426130-d9ee-11e6-848b-236d8f1f45df.png)

### Chart on browse page front end
![screen shot 2017-01-14 at 12 13 45 am](https://cloud.githubusercontent.com/assets/702526/21952563/713fcf42-d9ee-11e6-8d35-c233f12e7ff2.png)
![screen shot 2017-01-14 at 12 13 41 am](https://cloud.githubusercontent.com/assets/702526/21952565/714294de-d9ee-11e6-8c0d-2df66b55b03b.png)


## Todos

For this PR, my goals are to:

- [x] get the JavaScript for a line chart to render svgs for multiple charts on the page
- [x] use the data attributes output on the page HTML to target JS for each graph
- [x] whittle the admin fields for Chart Block so we only use those necessary for charts to render

## Checklist
